### PR TITLE
executor: do not acqurie pessimistic lock for non-unique index keys

### LINKED ISSUE #36235: Pessimistic DML should not lock non-unique index keys
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t(id varchar(40), v int, PRIMARY KEY (`id`) clustered, key i2(v))
insert into t values ('123', 1)
begin pessimistic
update t set v = v + 1 where id = '123'
commit
```

### 2. What did you expect to see? (Required)

The update statement is not supposed to lock the index key.

### 3. What did you see instead (Required)

The update statement acquires a pessimistic lock for the index key.

### 4. What is your TiDB version? (Required)

master (987bdd3014d8c90302caac3e99092d8014c47109)

### PR DESCRIPTION
**Pull request by ekexium**
Signed-off-by: ekexium <eke@fastmail.com>

### What problem does this PR solve?

Issue Number: close #36235 

Problem Summary:

Fix the condition of checking if the index KV is unique or not.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: session/txn.go [hunk 1]
```diff
@@ -488,27 +493,46 @@ func (txn *LazyTxn) Wait(ctx context.Context, sctx sessionctx.Context) (kv.Trans
...
 	if len(v) == 0 {
-		return flags.HasNeedLocked() || tablecodec.IsRecordKey(k)
+		return flags.HasNeedLocked() || tablecodec.IsRecordKey(k), nil
 	}
 
 	if tablecodec.IsUntouchedIndexKValue(k, v) {
-		return false
+		return false, nil
 	}
-	isNonUniqueIndex := tablecodec.IsIndexKey(k) && len(v) == 1
```
Comment (tiancaiamao) [inline session/txn.go:509]: Why the old code fail to work?
For unique index, the value is encoded in the kv key part, so kv value part is not set.
For non-unique index, the value is encoded in the kv value part.
Can't `len(v) == 1` distinguish them?

Comment (tiancaiamao) [reply session/txn.go:509]: OK, get it.
In a previous refactoring, the [key-value encoding format](https://docs.google.com/document/d/1Co5iMiaxitv3okJmLYLJxZYCNChcjzswJMRr-_45Eqg/edit#heading=h.l238zyi79ds4) is changed

image

**Comment by TonsnakeLin** [comment]
LGTM

**Review by TonsnakeLin — APPROVED** [review]

**Review by wjhuang2016 — COMMENTED** [review]
You can ref https://github.com/wjhuang2016/tidb/commit/58a4c2a440af2bcceed83e6a9a630cba3a9ee239#diff-9b9ca7ecfffe62c4dbeb06fe59c7c2172086aaa2cd06fb8370b8fd0b7a367cd7R910-R920

**Review by tangenta — APPROVED** [review]

File: session/txn.go (hunk 1 above)
Comment (TonsnakeLin) [reply session/txn.go:509]: In which version we changed key-value encoding format ?

Comment (tiancaiamao) [reply session/txn.go:509]: Since the introducing of global index / clustered index / collation, whichever comes first. @TonsnakeLin

**Review by tiancaiamao — APPROVED** [review]

File: session/txn.go [hunk 2]
```diff
@@ -506,9 +506,12 @@ func keyNeedToLock(k, v []byte, flags kv.KeyFlags) bool {
...
-	// Put row key and unique index need to lock.
-	return !isNonUniqueIndex
+
+	if !tablecodec.IsIndexKey(k) {
+		return true
+	}
+
+	return tablecodec.IndexKVIsUnique(v)
```
Comment (wjhuang2016) [inline session/txn.go:514]: So, we don't need to consider the untouched flag here?

Comment (ekexium) [reply session/txn.go:514]: It has checked `IsUntouchedIndexKValue` above.

Comment (wjhuang2016) [reply session/txn.go:514]: ok

**Pull request merged**